### PR TITLE
Initial implementation for TPL

### PR DESCRIPTION
**Pull request by knarkzel**
Added TPL struct, `new()`, `raw()` for convering into `ffi::TPLFile` struct and `open_tpl_from_memory(...)`. The code needs reviewing to make it work like it should, since I'm not too familiar with Rust and C ffi.

### CODE REVIEW

**Comment by knarkzel** [comment]
I see this might serve as a guide for someone wanting to proof check this PR: https://github.com/ProfElements/void/blob/main/examples/wii/src/main.rs#L122-L133.

**Comment by ProfElements** [comment]
The tpl file doesn't close.  (Which is a know issue in my code). Please close the file on drop.

File: src/tpl.rs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+enum FileType {
+    Disc = 0,
+    Memory = 1,
+}
+
+struct Tpl {
+    type_: FileType,
+    ntextures: u32,
+    texdesc: Vec<u8>,
+    tpl_file: ffi::FHANDLE,
```
Comment (ProfElements) [inline src/tpl.rs:18]: Why is this an FHANDLE instead of a libc::FILE or something similar?

File: src/tpl.rs [hunk 2]
```diff
@@ -14,30 +14,41 @@ enum FileType {
...
+            ffi::TPL_OpenTPLFromMemory(tpl.raw(), memory.as_mut_ptr() as *mut _, memory.len() as _)
+        };
+        tpl
     }
 
-    fn open_tpl_from_memory(&mut self, memory: &mut [u8]) {
-        unsafe { ffi::TPL_OpenTPLFromMemory(self.raw(), memory.as_mut_ptr() as *mut _, memory.len() as u32 )};
+    // Loads texture by id into texture
+    pub fn get_texture(&mut self, id: u32, texture: &mut gx::Texture) {
+        unsafe { ffi::TPL_GetTexture(self.raw(), id as _, texture.gxtexobj()) };
```
Comment (ProfElements) [inline src/tpl.rs:37]: You should be able to do `texture.into()` with the From impl?

**Review by ProfElements — CHANGES_REQUESTED** [review]
Just a few little bits and then you will be all set.

File: src/tpl.rs (hunk 2 above)
Comment (knarkzel) [reply src/tpl.rs:37]: I tried to do that, but it didn't work.
```rust
error[E0277]: the trait bound `*mut _gx_texobj: From<&mut Texture<'_>>` is not satisfied
  --> src/tpl.rs:37:67
   |
37 |         unsafe { ffi::TPL_GetTexture(self.raw(), id as _, texture.into()) };
   |                                                                   ^^^^ the trait `From<&mut Texture<'_>>` is not implemented for `*mut _gx_texobj`
```

Comment (ProfElements) [reply src/tpl.rs:37]: Ohhh I see can you do a &*texture.into()?

Comment (ProfElements) [reply src/tpl.rs:37]: I cant resolve this :/

**Review by ProfElements — APPROVED** [review]
LGTM

**Review by FrictionlessPortals — APPROVED** [review]
LGTM, thanks!

**Pull request merged**